proc/proc.go

```diff
@@ -180,92 +209,131 @@ func (dbp *Process) FindFileLocation(fileName string, lineno int) (uint64, error
 // Pass lineOffset == 0 and firstLine == false if you want the address for the function's entry point
 // Note that setting breakpoints at that address will cause surprising behavior:
 // https://github.com/derekparker/delve/issues/170
-func (dbp *Process) FindFunctionLocation(funcName string, firstLine bool, lineOffset int) (uint64, error) {
-	origfn := dbp.goSymTable.LookupFunc(funcName)
+func FindFunctionLocation(p *Process, funcName string, firstLine bool, lineOffset int) (uint64, error) {
+	origfn := p.Dwarf.LookupFunc(funcName)
 	if origfn == nil {
 		return 0, fmt.Errorf("Could not find function %s\n", funcName)
 	}
 
 	if firstLine {
-		return dbp.FirstPCAfterPrologue(origfn, false)
+		return FirstPCAfterPrologue(p, origfn, false)
 	} else if lineOffset > 0 {
-		filename, lineno, _ := dbp.goSymTable.PCToLine(origfn.Entry)
-		breakAddr, _, err := dbp.goSymTable.LineToPC(filename, lineno+lineOffset)
+		filename, lineno, _ := p.Dwarf.PCToLine(origfn.Entry)
+		breakAddr, _, err := p.Dwarf.LineToPC(filename, lineno+lineOffset)
 		return breakAddr, err
 	}
 
 	return origfn.Entry, nil
 }
 
-// CurrentLocation returns the location of the current thread.
-func (dbp *Process) CurrentLocation() (*Location, error) {
-	return dbp.CurrentThread.Location()
-}
-
-// RequestManualStop sets the `halt` flag and
+// Stop sets the `halt` flag and
 // sends SIGSTOP to all threads.
-func (dbp *Process) RequestManualStop() error {
-	if dbp.exited {
+func Stop(p *Process) error {
+	if p.exited {
 		return &ProcessExitedError{}
 	}
-	dbp.halt = true
-	return dbp.requestManualStop()
+	p.halt = true
+	return stop(p)
+}
+
+// Kill kills the target process.
+func Kill(p *Process) (err error) {
+	if p.exited {
+		return nil
+	}
+	log.WithField("pid", p.Pid).Info("killing process")
+	if err = kill(p); err != nil {
+		return errors.New("could not deliver signal " + err.Error())
+	}
+	_, err = Mourn(p)
+	return err
 }
 
 // SetBreakpoint sets a breakpoint at addr, and stores it in the process wide
 // break point table. Setting a break point must be thread specific due to
 // ptrace actions needing the thread to be in a signal-delivery-stop.
-func (dbp *Process) SetBreakpoint(addr uint64) (*Breakpoint, error) {
-	if dbp.exited {
-		return nil, &ProcessExitedError{}
-	}
-	return dbp.setBreakpoint(dbp.CurrentThread.ID, addr, false)
+func SetBreakpoint(p *Process, addr uint64) (*Breakpoint, error) {
+	return setBreakpoint(p, p.CurrentThread, addr, false, p.arch.BreakpointInstruction())
 }
 
 // SetTempBreakpoint sets a temp breakpoint. Used during 'next' operations.
-func (dbp *Process) SetTempBreakpoint(addr uint64) (*Breakpoint, error) {
-	return dbp.setBreakpoint(dbp.CurrentThread.ID, addr, true)
+func SetTempBreakpoint(p *Process, addr uint64) (*Breakpoint, error) {
+	return setBreakpoint(p, p.CurrentThread, addr, true, p.arch.BreakpointInstruction())
+}
+
+func setBreakpoint(p *Process, mem memoryReadWriter, addr uint64, temp bool, instr []byte) (*Breakpoint, error) {
```

Comment:
I'm not sure what's the point of passing memoryReadWriter and instr separately, I don't think there can be a single circumstance where it would be useful for those to not be p.CurrentThread and p.arch.BreakpointInstruction.